libgadget/fdm.c

```diff
@@ -0,0 +1,585 @@
+        for(k = 0; k < 3; k++){
+            gradsq += pow(FDMP(i).GradDensity[k],2);
+        }
+        FDMP(i).LapDensity -= gradsq/FDMP(i).Density;
+    }
+//     message(0,"deriv,ID=%lu,LapDensity=%g,gradsq=%g,density=%g",P[i].ID,FDMP(i).LapDensity,
+//             FDMP(i).GradDensity[0],FDMP(i).Density);
+}
+
+/*************************************************************************/
+typedef struct{
+    TreeWalkQueryBase base;
+    MyFloat Hsml;
+} TreeWalkQueryQP;
+
+typedef struct{
+    TreeWalkResultBase base;
+    MyFloat Acc[3];
+} TreeWalkResultQP;
+
+typedef struct {
+    TreeWalkNgbIterBase base;
+    DensityKernel kernel_i;
+} TreeWalkNgbIterQP;
+
+static void
+qp_copy(int place, TreeWalkQueryQP * I, TreeWalk * tw)
+{
+    I->Hsml = P[place].Hsml;
+}
+
+static void
+qp_reduce(int place, TreeWalkResultQP * remote, enum TreeWalkReduceMode mode, TreeWalk * tw)
+{
+    int k;
+    for(k=0;k<3;k++){
+        TREEWALK_REDUCE(FDMP(place).QPAccel[k], remote->Acc[k]);
+    }
+//     message(0,"Reduced,ID=%lu,QPAccel=(%g|%g|%g),GravAccel=(%g|%g|%g)",P[place].ID,FDMP(place).QPAccel[0],FDMP(place).QPAccel[1],
+//             FDMP(place).QPAccel[2],P[place].GravAccel[0],P[place].GravAccel[1],P[place].GravAccel[2]);
+}
+
+static void
+qp_ngbiter(
+        TreeWalkQueryQP * I,
+        TreeWalkResultQP * O,
+        TreeWalkNgbIterQP * iter,
+        LocalTreeWalk * lv)
+{
+    if(iter->base.other == -1) {
+        iter->base.Hsml = I->Hsml;
+        iter->base.mask = 2; /* dm-only */
+        iter->base.symmetric = NGB_TREEFIND_ASYMMETRIC;
+        
+        O->Acc[0] = O->Acc[1] = O->Acc[2] = 0;
+        density_kernel_init(&iter->kernel_i, I->Hsml, GetDensityKernelType());
+        return;
+    }
+    
+    const int other = iter->base.other;
+    const double r = iter->base.r;
+    const double r2 = iter->base.r2;
+    const double * dist = iter->base.dist;
+    
+    if(P[other].Type != 1 || r2 <=0 || !(r2 < iter->kernel_i.HH))
+        return;
+    
+    const double u = r * iter->kernel_i.Hinv;
+    const double dwk = density_kernel_dwk(&iter->kernel_i, u);
+    const double mass_j = P[other].Mass;
+    const double rho_j = FDMP(other).Density;
+    
+    double h_bar = PLANCK/(2*M_PI); /* in cgs unit */
+    double m_boson = FdmParams.FDM22*1.0e-22*eVinergs/(LIGHTCGS*LIGHTCGS); /*to cgs*/
+    double prefac = pow((h_bar/m_boson),2)/2.0;
+    /* convert to internal unit*/
+    prefac *= 1.0/pow((All.UnitVelocity_in_cm_per_s*All.UnitLength_in_cm),2);
+    
+    double gradsq = 0;
+    int d;
+    for(d = 0; d < 3; d++){
+        gradsq += pow(FDMP(other).GradDensity[d],2);
+    }
+    double fac = FDMP(other).LapDensity/(2*rho_j) - gradsq/(4*rho_j*rho_j);
+    fac *= prefac*(mass_j/rho_j)*FDMP(other).DhsmlDensityFactor;
+    /* Convention of QP_acc unit: same as Grav_Accel, that [Acc_code] = [Acc_physical]*a*a 
+     * As QP is the third order derivative of distance, so [Acc_code] = (a3inv*QP)*a*a = QP/a
+     */
+    fac /= (All.cf.a);
```

Comment:
Need double check here, did I get the a_scale factor correct?
I am a bit confused why https://arxiv.org/pdf/2101.01828.pdf Eq.13 following https://arxiv.org/pdf/1911.05746.pdf Eq.9 has a^-2 (instead of a^-3) before the QP term.

Reply:
With an extra 1/a factor for QP, the effect on suppression of ps matches Nori's result surprisingly well from z=49 to z=3:
<img width="383" alt="ps-compare" src="https://user-images.githubusercontent.com/52514409/113163459-31f0e380-920e-11eb-8721-d7bcd29804f0.png">